hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/CassAlertsServiceImpl.java

```diff
@@ -0,0 +1,849 @@
+                if (triggerIds.size() == 1) {
+                    sAlertsTriggers.append("( triggerId = '");
+                    sAlertsTriggers.append(triggerIds.iterator().next());
+                    sAlertsTriggers.append("' )");
+                } else {
+                    sAlertsTriggers.append("( triggerId IN (");
+                    int entries = 0;
+                    for (String triggerId : triggerIds) {
+                        if (isEmpty(triggerId)) {
+                            continue;
+                        }
+                        sAlertsTriggers.append(entries++ > 0 ? "," : "");
+                        sAlertsTriggers.append("'");
+                        sAlertsTriggers.append(triggerId);
+                        sAlertsTriggers.append("'");
+                    }
+                    sAlertsTriggers.append(") )");
+                }
+                try {
+                    ResultSet rsAlertIdsByTriggerIds = session.execute(sAlertsTriggers.toString());
+                    if (rsAlertIdsByTriggerIds.isExhausted()) {
+                        alertIdsFilteredByTriggers.add("empty-alert");
+                    } else {
+                        Iterator<Row> itAlertIdsByTriggerIds = rsAlertIdsByTriggerIds.iterator();
+                        while (itAlertIdsByTriggerIds.hasNext()) {
+                            Row row = itAlertIdsByTriggerIds.next();
+                            String alertId = row.getString("alertId");
+                            alertIdsFilteredByTriggers.add(alertId);
+                        }
+                    }
+                } catch (Exception e) {
+                    msgLog.errorDatabaseException(e.getMessage());
+                    throw e;
+                }
+            }
+
+            /*
+                Add ctime clause
+             */
+            Set<String> alertIdsFilteredByCtime = new HashSet<>();
+            boolean filterByCtime = false;
+            if (criteria.getStartTime() != null || criteria.getEndTime() != null) {
+                filterByCtime = true;
+                StringBuilder sAlertsCtimes = new StringBuilder("SELECT alertId FROM ").append(keyspace)
+                        .append(".alerts_ctimes WHERE ");
+                boolean andNeeded = false;
+                if (criteria.getStartTime() != null) {
+                    sAlertsCtimes.append("( ctime >= ");
+                    sAlertsCtimes.append(criteria.getStartTime());
+                    sAlertsCtimes.append(" )");
+                    andNeeded = true;
+                }
+                if (criteria.getEndTime() != null) {
+                    if (andNeeded) {
+                        sAlertsCtimes.append(" AND ");
+                    }
+                    sAlertsCtimes.append("( ctime <= ");
+                    sAlertsCtimes.append(criteria.getEndTime());
+                    sAlertsCtimes.append(" )");
+                }
+                sAlertsCtimes.append(" ALLOW FILTERING ");
```

Comment:
I absolutely agree with merging changes sooner rather than later. The best way to learn is to have people beat on it, and that won't happen until the changes get merged. With that said, a lot of the things I have mentioned are not simply matters of scale. They are recommended/best practices and anti-patterns. Moving to Cassandra is a big paradigm shift, and Lucas is off to a really nice start; however, you want to cognizant of things you should and should not do regardless of scale.